tests/cluster/etl/test_extended_transformer.py

```diff
@@ -0,0 +1,90 @@
+
+from atc.etl import ExtendedTransformer
+from atc.etl.types import dataset_group
+from atc.spark import Spark
+
+
+class TransformerTests(unittest.TestCase):
+    @classmethod
+    def setUpClass(cls):
+
+        cls.transformer = TestTransformer()
+        cls.df = create_dataframe()
+
+        cls.transformer_output_key = TestTransformer(dataset_output_key="my_output_key")
+        cls.transformer_input_key = TestTransformer(dataset_input_key="my_input_key")
+        cls.transformer_input_key_list = TestTransformer(
+            dataset_input_key_list=["my_input_key_1", "my_input_key_2"]
+        )
+
+    def test_process(self):
+        result = self.transformer.etl({"df": self.df})
+        self.assertIs(self.df, list(result.values())[1])
+
+    def test_process_many(self):
+        result = self.transformer.etl({"df1": self.df, "df2": self.df})
+        self.assertEqual(6, list(result.values())[2].count())
+
+    def test_process_no_key(self):
+        result = self.transformer.etl({"df": self.df})
+        self.assertEqual("TestTransformer", list(result.keys())[1])
+
+    def test_process_many_no_key(self):
+        result = self.transformer.etl({"df1": self.df, "df2": self.df})
+        self.assertEqual("TestTransformer", list(result.keys())[2])
+
+    def test_process_output_key(self):
+        result = self.transformer_output_key.etl({"df": self.df})
+        self.assertEqual("my_output_key", list(result.keys())[1])
+
+    def test_process_many_output_key(self):
+        result = self.transformer_output_key.etl({"df1": self.df, "df2": self.df})
+        self.assertEqual("my_output_key", list(result.keys())[2])
+
+    def test_process_input_key(self):
+        result = self.transformer_input_key.etl({"my_input_key": self.df})
+        self.assertIs(self.df, list(result.values())[1])
+
+        with self.assertRaises(KeyError):
+            self.transformer_input_key.etl({"df": self.df})
+
+    def test_process_many_input_key_list(self):
+        result = self.transformer_input_key_list.etl(
+            {"my_input_key_1": self.df, "my_input_key_2": self.df}
+        )
+        self.assertEqual(6, list(result.values())[2].count())
+
+        with self.assertRaises(AssertionError):
+            self.transformer_input_key_list.etl({"df1": self.df, "df2": self.df})
+
+
+class TestTransformer(ExtendedTransformer):
+    def process(self, df: DataFrame) -> DataFrame:
+        return df
+
+    def process_many(self, datasets: dataset_group) -> DataFrame:
+        assert len(datasets) == 2
+        df1, df2 = list(datasets.values())
+        return df1.union(df2)
+
+
+def create_dataframe():
+
+    data = [(1, "1"), (2, "2"), (3, "3")]
+
+    schema = StructType(
+        [
+            StructField("id", IntegerType(), False),
+            StructField("text", StringType(), False),
+        ]
+    )
+
+    return Spark.get().createDataFrame(data=data, schema=schema)
+
+
+if __name__ == "__main__":
```

Comment:
why do you need line 89-90?

Reply:
I ran these locally when writing the tests to check along the way. I just left it for ease in the future - should remove?

Reply:
In my opinion I would remove it.